From 19b98f31459c5412a348f51e813b637ae0dac60e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:13:40 -0400
Subject: [powerpc] bpf: Use correct mask while accessing the VLAN tag

Message-id: <1410545655-205645-592-git-send-email-dzickus@redhat.com>
Patchwork-id: 94409
O-Subject: [RHEL7 PATCH 591/626] powerpc: bpf: Use correct mask while accessing the VLAN tag
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 3fc60aa097b8eb0f701c5bf755bc8f7d3ffeb0bd
Author: Denis Kirjanov <kda@linux-powerpc.org>
Date:   Wed Jun 25 21:34:56 2014 +0400

    powerpc: bpf: Use correct mask while accessing the VLAN tag

    To get a full tag (and not just a VID) we should access the TCI
    except the VLAN_TAG_PRESENT field (which means that 802.1q header
    is present). Also ensure that the VLAN_TAG_PRESENT stay on its place

    Signed-off-by: Denis Kirjanov <kda@linux-powerpc.org>
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/net/bpf_jit_comp.c b/arch/powerpc/net/bpf_jit_comp.c
index b5573a8..efea11d 100644
--- a/arch/powerpc/net/bpf_jit_comp.c
+++ b/arch/powerpc/net/bpf_jit_comp.c
@@ -396,10 +396,12 @@ static int bpf_jit_build_body(struct sk_filter *fp, u32 *image,
   case BPF_S_ANC_VLAN_TAG:
   case BPF_S_ANC_VLAN_TAG_PRESENT:
    BUILD_BUG_ON(FIELD_SIZEOF(struct sk_buff, vlan_tci) != 2);
+   BUILD_BUG_ON(VLAN_TAG_PRESENT != 0x1000);
+
    PPC_LHZ_OFFS(r_A, r_skb, offsetof(struct sk_buff,
          vlan_tci));
    if (filter[i].code == BPF_S_ANC_VLAN_TAG)
-    PPC_ANDI(r_A, r_A, VLAN_VID_MASK);
+    PPC_ANDI(r_A, r_A, ~VLAN_TAG_PRESENT);
    else
     PPC_ANDI(r_A, r_A, VLAN_TAG_PRESENT);
    break;
-- 
1.7.1